### ISSUE TITLE
Differences between Matlab and Python reconstructions using ASTRA Toolbox

### ISSUE THREAD

**Issue by vribgar** [astra-toolbox/astra-toolbox#448]
We have made a code in Matlab, using the ASTRA package, which consists of taking projections of an object (in this case a volume representing the human torso) and from these projections, obtain a reconstruction. We wanted to reproduce this same code in Python, but in doing so we have seen some differences in the reconstructions obtained.

In the attached image, the Volume Viewer window showing the variable "rec_rot" is the reconstruction obtained with Python, while "rec_v" is the reconstruction obtained in Matlab. Clear differences can be seen in terms of sharpness and contrast of the structures. Para visualizar las reconstrucciones obtenidas en Python en el Volume Viewer de Matlab se ha utilizado npy-matlab (readNPY).
![Python_Matlab_comparison](https://github.com/astra-toolbox/astra-toolbox/assets/150797747/80d84702-48c5-421f-a3f0-24d6976b784f)

The PSNR of the two reconstructions obtained with the initial volume was checked and it turned out that the PSNR of the Matlab reconstruction was higher than that of the Python reconstruction (attached image).
![psnr](https://github.com/astra-toolbox/astra-toolbox/assets/150797747/65cf2cee-5e88-4ab8-9620-0803bc8a4237)

In addition, it has been observed that the Matlab reconstruction is saved as "double" and in Python as "single".

Is there any way to get the same reconstruction quality as in Matlab using Python?

Thank you very much for your attention.

**Comment by askorikov** [comment]
Hi! That's quite unexpected, because both Matlab and Python interfaces call the same underlying C++ implementation of algorithms. Could you maybe share the code you used to do the reconstruction in Matlab and Python?

**Comment by vribgar** [comment]
Hi, thanks for your quick reply, here are the codes! But I can't pass you the volume (C012_full_att.mat) as it is too big (340 MB).
[Codes UPV.zip](https://github.com/astra-toolbox/astra-toolbox/files/14257783/Codes.UPV.zip)

**Comment by askorikov** [comment]
I think it may be related to the fact that in Matlab version the input is rescaled to [dim1, dim2, dim1] = [ndetx, ndety, ndetx] = [243, 307, 243], whereas in Python these factors are hard-coded to [243, 243, 243]. Therefore, in the Python case you lose some information along the Y  dimension.

**Comment by vribgar** [comment]
The object to be reconstructed (C012_full_att) has dimensions 512x512x351 and in the variable p the resized object (243x243x243) is stored both in the Matlab and Python code... I don't know if that's what you mean...

Thanks for your answer!

**Comment by askorikov** [comment]
Oh, you're right. `dim2` is set to be equal to `dim1` later in the script. Well, I'm not sure why the reconstructions are different then. I will try to run both scripts on synthetic data later, and in the meantime I would suggest double-checking that all geometric parameters are the same.

P.S. Both in Python and Matlab computations are done in "single" precision. In the Matlab case it is just converted to "double" after the fact , so that is not likely to be the source of the difference.

**Comment by vribgar** [comment]
In the data3d_c file (https://github.com/astra-toolbox/astra-toolbox/blob/master/python/astra/data3d_c.pyx) at line 198 (function get(i)): 

def get(i):
    cdef CFloat32Data3DMemory * pDataObject = dynamic_cast_mem_safe(getObject(i))
    outArr = np.empty((pDataObject.getDepth(),pDataObject.getHeight(),pDataObject.getWidth()),dtype=np.float32,order='C')
    cdef float [:,:,::1] mView = outArr
    cdef float [:,:,::1] cView = <float[:outArr.shape[0],:outArr.shape[1],:outArr.shape[2]]> pDataObject.getData()
    mView[:] = cView
    return outArr

When creating the output array (outArr) in that get it passes dtype=np.float32 instead of reading the data type returned by the C code. 
The complete sequence of operations in the code begins with the invocation of the create_sino3d_gpu(...) function on line 452 of the creators.py file (https://github.com/astra-toolbox/astra-toolbox/blob/master/python/astra/creators.py). During this process, the function data3d.get(sino_id) is executed for the purpose of obtaining the sinogram data. This step is part of the interface that, in turn, activates the get function of data3d_c, where the error is found.

**Comment by askorikov** [comment]
I'm not completely sure what you mean by that, but this is basically what I said. The result of the computation (`cView`) has `float` (= `single`, = `np.float32`) datatype. In Matlab this result is just converted into `double` datatype after the fact by default.

**Comment by vribgar** [comment]
Yes! You're right... if you want to use the volume that we have used for the reconstructions to make some tests I can send it to you by mail.

Thank you again.

**Comment by vribgar** [comment]
Hi! we have been checking and it seems that although the resizing of the initial volume that is stored in the variable "p" is visually similar in Matlab and Python, its characteristics differ a bit, a fact that may contribute to different reconstructions. 
It is quite possible that this is the problem and not something related to the operation of ASTRA Toolbox. 

Thank you very much for your attention and sorry for the inconvenience!

**Comment by askorikov** [comment]
Good point! You can try [scikit-image](https://scikit-image.org/docs/stable/api/skimage.transform.html#skimage.transform.resize) and maybe OpenCV implementations of resizing to see if it changes the outcome.
